hir/src/asm.rs

```diff
@@ -0,0 +1,2447 @@
+            MasmOp::U32Not => f.write_str("u32.not"),
+            MasmOp::U32CheckedShl => f.write_str("u32.shl.checked"),
+            MasmOp::U32CheckedShlImm(imm) => write!(f, "u32.shl.checked.{}", imm),
+            MasmOp::U32UncheckedShl => f.write_str("u32.shl.unchecked"),
+            MasmOp::U32UncheckedShlImm(imm) => write!(f, "u32.shl.unchecked.{}", imm),
+            MasmOp::U32CheckedShr => f.write_str("u32.shr.checked"),
+            MasmOp::U32CheckedShrImm(imm) => write!(f, "u32.shr.checked.{}", imm),
+            MasmOp::U32UncheckedShr => f.write_str("u32.shr.unchecked"),
+            MasmOp::U32UncheckedShrImm(imm) => write!(f, "u32.shr.unchecked.{}", imm),
+            MasmOp::U32CheckedRotl => f.write_str("u32.rotl.checked"),
+            MasmOp::U32CheckedRotlImm(imm) => write!(f, "u32.rotl.checked.{}", imm),
+            MasmOp::U32UncheckedRotl => f.write_str("u32.rotl.unchecked"),
+            MasmOp::U32UncheckedRotlImm(imm) => write!(f, "u32.rotl.unchecked.{}", imm),
+            MasmOp::U32CheckedRotr => f.write_str("u32.rotr.checked"),
+            MasmOp::U32CheckedRotrImm(imm) => write!(f, "u32.rotr.checked.{}", imm),
+            MasmOp::U32UncheckedRotr => f.write_str("u32.rotr.unchecked"),
+            MasmOp::U32UncheckedRotrImm(imm) => write!(f, "u32.rotr.unchecked.{}", imm),
+            MasmOp::U32CheckedPopcnt => f.write_str("u32.popcnt.checked"),
+            MasmOp::U32UncheckedPopcnt => f.write_str("u32.popcnt.unchecked"),
+            MasmOp::U32Eq => f.write_str("u32.eq"),
+            MasmOp::U32EqImm(imm) => write!(f, "u32.eq.{:#0x}", imm),
+            MasmOp::U32Neq => f.write_str("u32.neq"),
+            MasmOp::U32NeqImm(imm) => write!(f, "u32.neq.{:#0x}", imm),
+            MasmOp::U32CheckedLt => f.write_str("u32.lt.checked"),
+            MasmOp::U32UncheckedLt => f.write_str("u32.lt.unchecked"),
+            MasmOp::U32CheckedLte => f.write_str("u32.lte.checked"),
+            MasmOp::U32UncheckedLte => f.write_str("u32.lte.unchecked"),
+            MasmOp::U32CheckedGt => f.write_str("u32.gt.checked"),
+            MasmOp::U32UncheckedGt => f.write_str("u32.gt.unchecked"),
+            MasmOp::U32CheckedGte => f.write_str("u32.gte.checked"),
+            MasmOp::U32UncheckedGte => f.write_str("u32.gte.unchecked"),
+            MasmOp::U32CheckedMin => f.write_str("u32.min.checked"),
+            MasmOp::U32UncheckedMin => f.write_str("u32.min.unchecked"),
+            MasmOp::U32CheckedMax => f.write_str("u32.max.checked"),
+            MasmOp::U32UncheckedMax => f.write_str("u32.max.unchecked"),
+        }
+    }
+}
+
+pub trait StackElement: Clone + fmt::Debug {
+    /// A value of this type which represents the "zero" value for the type
+    const DEFAULT: Self;
+}
+impl StackElement for Felt {
+    const DEFAULT: Self = Felt::ZERO;
+}
+impl StackElement for Type {
+    const DEFAULT: Self = Type::Felt;
+}
+
+pub trait Stack: std::ops::IndexMut<usize, Output = <Self as Stack>::Element> {
```

Comment:
There's an analysis going on there, but it's not clear to me what the analysis is. Is it just keeping track of the stack height?

Reply:
This trait isn't doing any particular kind of analysis, aside from just generally representing what happens to the operand stack when the stack manipulation ops are applied. It is up to a specific implementation of this trait to define the specific behavior though.

That said, I'll add some documentation to the trait to describe it's default behavior. It is primarily meant to take care of the common plumbing of the operand stack, but some ops, like `cswap` and such, are more sensitive to the specific behavior desired. I may remove the default implementation for those, or move them to the concrete implementation(s), since they are less general.